### PR TITLE
CM-11 unsupportedConfigOverrides for the operator

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.redhat.com/projects/CM/issues/CM-11

This Pull Request introduces the support for `unsupportedConfigOverrides`. Note there are no tests for this - it is unsupported functionality (a sort of backdoor to inject custom settings if needed).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -117,3 +117,23 @@ Then, let it run for a few minutes. Once the operands are deployed, just invoke:
 
 The `-o test-e2e-wait-for-stable-state` skips checking if the operator is deployed. Depending on how do you're testing it,
 you may or may not want to invoke it.
+
+## Using unsupported config overrides options
+
+It is possible (although not supported) to specify custom settings to each Cert Manager image. In order to do it, 
+you need to modify the `co/certmanager` object:
```
Comment (xingxingxia) [inline README.md:124]:
```
you need to modify the `co/certmanager` object
```
Maybe below is preciser:
``
you need to modify the `certmanager.operator/cluster` object
```

**Comment by s-urbaniak** [comment]
> Note there are no tests for this - it is unsupported functionality (a sort of backdoor to inject custom settings if needed).

despite being an unsupported functionality we should have a test.

File: README.md (hunk 1 above)
Comment (slaskawi) [reply README.md:124]: :+1:

**Comment by slaskawi** [comment]
@s-urbaniak Added a test.

File: pkg/controller/deployment/generic_deployment_controller.go [hunk 2]
```diff
@@ -80,7 +92,14 @@ func (c *genericDeploymentController) Sync(ctx context.Context, syncContext fact
...
 		deployment.Spec.Template.Spec.Containers[index].Image = certManagerImage(deployment.Spec.Template.Spec.Containers[index].Image)
 	}
-	_, opStatus, _, _ := c.operatorClient.GetOperatorState()
+	operatorSpec, opStatus, _, _ := c.operatorClient.GetOperatorState()
+	unsupportedExtensions, err := operatorclient.GetUnsupportedConfigOverrides(operatorSpec)
+	if err != nil {
+		return nil, false, append(errors, fmt.Errorf("Getting unsupportedConfigOverrides failed: %w", err))
+	}
+
+	deployment = c.customOverridesFunc(deployment, unsupportedExtensions)
```
Comment (s-urbaniak) [inline pkg/controller/deployment/generic_deployment_controller.go:101]: unless this is dictated by library-go, why would we need to the abstraction via a custom overrides func, why not executing it directly in the Sync method? The function could still be tested in isolation.

Comment (slaskawi) [reply pkg/controller/deployment/generic_deployment_controller.go:101]: aaaahh.. it's a leftover from previous approach. Good catch, let me refactor it.

File: pkg/controller/deployment/custom_overrides_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+			},
+			wantArgs: defaultArgs,
+		},
+		"Controller overrides should work": {
+			overrides: &v1alpha1.UnsupportedConfigOverrides{
+				Controller: v1alpha1.UnsupportedConfigOverridesForCertManagerController{
+					Args: []string{"test"},
+				},
+			},
+			wantArgs: append(defaultArgs, "test"),
```
Comment (slaskawi) [inline pkg/controller/deployment/custom_overrides_test.go:45]: I just noticed, this is absolutely wrong!!!

We should be overriding settings not appending new ones.

File: pkg/controller/deployment/generic_deployment_controller.go (hunk 2 above)
Comment (slaskawi) [reply pkg/controller/deployment/generic_deployment_controller.go]: Fixed.

File: pkg/controller/deployment/custom_overrides.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package deployment
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+
+	"github.com/openshift/cert-manager-operator/apis/operator/v1alpha1"
+)
+
+func CustomOverrides(deployment *appsv1.Deployment, customOverrides *v1alpha1.UnsupportedConfigOverrides) *appsv1.Deployment {
```
Comment (s-urbaniak) [inline pkg/controller/deployment/custom_overrides.go:9]: s/CustomOverrides/UnsupportedConfigOverrides

File: pkg/controller/deployment/custom_overrides_test.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	"testing"
+
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/stretchr/testify/require"
+
+	"github.com/openshift/cert-manager-operator/apis/operator/v1alpha1"
+	"github.com/openshift/cert-manager-operator/pkg/operator/assets"
+)
+
+func TestCustomOverrides(t *testing.T) {
```
Comment (s-urbaniak) [inline pkg/controller/deployment/custom_overrides_test.go:13]: same as above

File: pkg/controller/deployment/custom_overrides.go (hunk 4 above)
Comment (slaskawi) [reply pkg/controller/deployment/custom_overrides.go]: Done

File: pkg/controller/deployment/custom_overrides_test.go (hunk 5 above)
Comment (slaskawi) [reply pkg/controller/deployment/custom_overrides_test.go]: Done

**Comment by slaskawi** [comment]
@s-urbaniak Ready for the next round of reviews.

**Pull request merged**